test/e2e/aks_public_ip_prefix.go

```diff
@@ -77,9 +77,9 @@ func AKSPublicIPPrefixSpec(ctx context.Context, inputGetter func() AKSPublicIPPr
 	})
 	Expect(err).NotTo(HaveOccurred())
 	var publicIPPrefix network.PublicIPPrefix
-	Eventually(func() error {
+	Eventually(func(g Gomega) {
 		publicIPPrefix, err = publicIPPrefixFuture.Result(publicIPPrefixClient)
-		return err
+		g.Expect(err).NotTo(HaveOccurred())
```

Comment:
do we actually need the `g.` since gomega is imported as `. "github.com/onsi/gomega"` ? everywhere else we just do `Expect()`, is this different because it's withing that func?

Reply:
Yes, at least that's how Gomega's example looks: https://onsi.github.io/gomega/#category-3-making-assertions-eminem-the-function-passed-into-codeeventuallycode

I haven't actually tried it, but I suspect when a bare `Expect` fails in an `Eventually`, then the whole test will immediately fail, but when the local `g.Expect` fails, then just that iteration of `Eventually` fails and it gets retried.

Reply:
This is my experience as well.

My main question at this point is when we use this pattern, are the individual `g.Expect(err).NotTo(HaveOccurred())` assertions going to bubble up to stdout/stderr? If we don't get that resolution (lots of these `Eventually` blocks are transactions with many possible, and very different, retryable error outcomes. We want to see exactly what is failing in these retry blocks.

Reply:
My query was answered in the affirmative. Just for fun I updated an existing `Eventually` block to this and then ran local E2E:

```golang
	Eventually(func(g Gomega) {
		err := errors.New("I am an error")
		g.Expect(err).NotTo(HaveOccurred())
		err = mgmtClient.Get(ctx, key, kubeadmControlPlane)
		g.Expect(err).NotTo(HaveOccurred())
	}, e2eConfig.GetIntervals(specName, "wait-daemonset")...).Should(Succeed(), "Failed to get KubeadmControlPlane object %s/%s", cluster.Spec.ControlPlaneRef.Namespace, cluster.Spec.ControlPlaneRef.Name)
```

Got:

```
  [FAILED] Timed out after 900.002s.
  Failed to get KubeadmControlPlane object capz-e2e-56ic9a/capz-e2e-56ic9a-ipv6-control-plane
  The function passed to Eventually failed at /Users/jackfrancis/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/daemonsets.go:69 with:
  Unexpected error:
      <*errors.errorString | 0xc0001d6050>: {s: "I am an error"}
      I am an error
  occurred
  In [It] at: /Users/jackfrancis/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/daemonsets.go:72 @ 01/26/23 12:01:53.315

  Full Stack Trace
    sigs.k8s.io/cluster-api-provider-azure/test/e2e.EnsureDaemonsets({0x6eefa60?, 0xc00019c010}, 0xc000573680)
    	/Users/jackfrancis/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/daemonsets.go:72 +0x783
    sigs.k8s.io/cluster-api-provider-azure/test/e2e.glob..func1.6.1.1()
    	/Users/jackfrancis/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/azure_test.go:372 +0x53
    sigs.k8s.io/cluster-api/test/framework/clusterctl.ApplyClusterTemplateAndWait({_, _}, {{0x6f02050, 0xc0007f8f40}, {{0xc000765800, 0x5e}, {0xc0018f804c, 0x6d}, {0xc0018faa9e, 0x42}, ...}, ...}, ...)
    	/Users/jackfrancis/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.3.2/framework/clusterctl/clusterctl_helpers.go:364 +0x138f
    sigs.k8s.io/cluster-api-provider-azure/test/e2e.glob..func1.6.1()
    	/Users/jackfrancis/go/src/sigs.k8s.io/cluster-api-provider-azure/test/e2e/azure_test.go:351 +0x4fe
```